<launch>
  <arg name="arm" default="left"/>

  <arg name="minX" default="-1.0"/>
  <arg name="maxX" default="1.0"/>
  <arg name="minY" default="-1.0"/>
  <arg name="maxY" default="1.0"/>
  <arg name="minZ" default="-1.0"/>
  <arg name="maxZ" default="1.0"/>
  <arg name="leaf_size" default="0.01"/>

  <arg name="distance_threshold" default="0.04"/>
  <arg name="cluster_tolerance" default="0.03"/>
  <arg name="min_cluster_size" default="30"/>
  <arg name="max_cluster_size" default="500"/>



  <arg name="crop_input_cloud"      default="/$(arg arm)_camera/depth_registered/points"/>
  <arg name="down_input_cloud"      default="/$(arg arm)_camera/depth_registered/points_crop"/>
  <arg name="down_output_cloud"     default="/$(arg arm)_camera/depth_registered/downsampled_points"/>
  <arg name="filtered_output_cloud" default="/$(arg arm)_camera/depth_registered/points_filtered"/>
  <arg name="output_cloud"          default="/$(arg arm)_camera/depth_registered/object_clusters"/>  
  <arg name="debug_output_cloud"    default="/$(arg arm)_camera/depth_registered/objects"/>

  <!-- launch the nodelet manager -->
  <node pkg="nodelet" type="nodelet" name="pcl_manager" args="manager" output="screen"/>  
  <node pkg="nodelet" type="nodelet" name="cropbox" args="load pcl/CropBox /pcl_manager" output="screen">
    <remap from="~input" to="$(arg crop_input_cloud)"/>
    <remap from="~output" to="$(arg down_input_cloud)"/>
    <param name="~output_frame" 		value="world"/>
    <param name="min_x"		value="$(arg minX)"/>
    <param name="max_x"		value="$(arg maxX)"/>
    <param name="min_y"		value="$(arg minY)"/>
    <param name="max_y"		value="$(arg maxY)"/>
    <param name="min_z"		value="$(arg minZ)"/>
    <param name="max_z"		value="$(arg maxZ)"/>
  </node>
  <node pkg="nodelet" type="nodelet" name="voxel" args="load pcl/VoxelGrid /pcl_manager" output="screen">
    <remap from="~input" to="$(arg down_input_cloud)"/>
    <remap from="~output" to="$(arg down_output_cloud)"/>
    <param name="leaf_size"		value="$(arg leaf_size)"/>
    <param name="filter_field_name"		value="z"/>
    <param name="filter_limit_min"		value="-10.0"/>
    <param name="filter_limit_max"		value="10.0"/>
  </node>

  <node name="$(arg arm)_rgbd_filter" pkg="sensor_filters" type="pointcloud2_filter_chain" output="screen">
    <rosparam command="load" file="$(find toy_assembly)/config/left_camera_filter_real.yaml" />
    <remap from="~input" to="$(arg down_output_cloud)" />
    <remap from="~output" to="$(arg filtered_output_cloud)" />
  </node>

  <node name= "object_segmentation" pkg= "obj_segmentation" type= "segmentation" output="screen" respawn= "false">
    <param name="distance_threshold" value="$(arg distance_threshold)"/>
    <param name="cluster_tolerance"	 value="$(arg cluster_tolerance)"/>
    <param name="min_cluster_size"	 value="$(arg min_cluster_size)"/>
    <param name="max_cluster_size"	 value="$(arg max_cluster_size)"/>
    <param name="input_cloud" 		   value="$(arg filtered_output_cloud)"/>
    <param name="output_cloud" 		   value="$(arg output_cloud)"/>
    <param name="debug_output_cloud" value="$(arg debug_output_cloud)"/>
  </node>
  
  <node name="left_overlay_image" pkg="toy_assembly" type="overlay_image.py" output="screen" >  
    <param name="real" value="True"/>
    <param name="arm" value="left"/>
    <param name="obj_cluster_topic" 		value="$(arg output_cloud)"/>
  </node>  

</launch>